File: No_0020_Valid Parentheses/src/by_stack.go
/*

Description:

Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true



Example 2:

Input: "()[]{}"
Output: true



Example 3:

Input: "(]"
Output: false



Example 4:

Input: "([)]"
Output: false



Example 5:

Input: "{[]}"
Output: true

*/

package No_0020

import (
	"container/list"
	"sync"
)

func isValid(s string) bool {

	stack := newStack()

	pair := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	for _, char := range s {

		if leftBracket, exist := pair[char]; exist {
			// check when we mmet right parenthesis

			if stack.Empty() || (!stack.Empty() && stack.Peek() != leftBracket) {

				// parenthesis mismatch
				return false
			} else {
				// parenthesis match
				stack.Pop()
			}
		} else {
			// push left parenthesis into stack
			stack.Push(char)
		}
	}

	// Accept in all parenthesis pair are matched
	return stack.Empty()
}


// n : the length of input string s

//// Time Complexity: O( n )
//
// The overhead is the for loop to linear scan each character in s


//// Space Complexity: O( n )
//
// The overhead is the variable to keep a stack to record the left braces including '(', '[' and '{'
// The size of stack is at most O( N )



// Implement Stack by golang linked list

type Stack struct {
	list     *list.List
	safeLock *sync.RWMutex
}

func newStack() *Stack {

	list := list.New()
	safeLock := &sync.RWMutex{}

	return &Stack{list, safeLock}
}

func (stack *Stack) Push(element interface{}) {

	stack.safeLock.Lock()
	defer stack.safeLock.Unlock()

	stack.list.PushBack(element)

	return
}

func (stack *Stack) Pop() interface{} {

	stack.safeLock.Lock()
	defer stack.safeLock.Unlock()

	element := stack.list.Back()

	if element != nil {
		stack.list.Remove(element)
		return element.Value
	}
	return nil
}

func (stack *Stack) Peek() interface{} {

	stack.safeLock.Lock()
	defer stack.safeLock.Unlock()

	element := stack.list.Back()

	if element != nil {
		return element.Value
	}
	return nil
}

func (stack *Stack) Len() int {
	return stack.list.Len()
}

func (stack *Stack) Empty() bool {
	return stack.list.Len() == 0
}

File: No_0020_Valid Parentheses/by_stack.py
'''

Description:

Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true



Example 2:

Input: "()[]{}"
Output: true



Example 3:

Input: "(]"
Output: false



Example 4:

Input: "([)]"
Output: false



Example 5:

Input: "{[]}"
Output: true

'''


class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = []
        
		## dictionary
		# key: right bracket
		# value: corresponding left bracket
        pair = {
                ')':'(',
                ']':'[',
                '}':'{',
                }
        
        for char in s:
            
            if char in pair:
                # check when we meet right parenthesis
                
                if (not stack) or ( stack and (stack[-1] != pair[char]) ):
                
                    # parenthesis mismatch
                    return False
                
                else:
                    
                    # parenthesis match
                    stack.pop()
            
            else:
                # push left parenthesis into stack
                stack.append( char )
        
        
        # Accept if all parenthesis pair are matched.
        return not stack


# n : the length of input string s

## Time Complexity: O( n )
#
# The overhead is the for loop to linear scan each character in s


## Space Complexity: O( n )
#
# The overhead is the variable to keep a stack to record the left braces including '(', '[' and '{'
# The size of stack is at most O( N )


import unittest

class Testing( unittest.TestCase ):

    def test_case_1( self ):

        result = Solution().isValid( s="()" )
        self.assertEqual(result, True)


    def test_case_2( self ):

        result = Solution().isValid( s="()[]{}" )
        self.assertEqual(result, True)


    def test_case_3( self ):

        result = Solution().isValid( s="(]" )
        self.assertEqual(result, False)


    def test_case_4( self ):

        result = Solution().isValid( s="([)]" )
        self.assertEqual(result, False)


    def test_case_5( self ):

        result = Solution().isValid( s="{[]}" )
        self.assertEqual(result, True)



# entry point of main
if __name__ == '__main__':

    unittest.main()